### PR TITLE
Add direct conversion from spec.Schema to SMD schema.Schema

### PR DESCRIPTION
**Pull request by alexzielenski**
Intended to remove very expensive conversion via json & gnostic from apiserver.

### CODE REVIEW

File: pkg/schemaconv/openapi.go [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+			a.Map.ElementType = schema.TypeRef{}
+		}
+
+		if mapRelationship, err := getMapElementRelationship(m.Extensions); err != nil {
+			c.reportError(err.Error())
+		} else if len(mapRelationship) > 0 {
+			a.Map.ElementRelationship = mapRelationship
+		}
+
+	case "array":
```
Comment (apelisse) [inline pkg/schemaconv/openapi.go:180]: I don't really see anything incorrect here either.

**Review by apelisse — COMMENTED** [review]
I haven't reviewed everything yet, clearly but I've left some comments.

File: pkg/schemaconv/openapi.go [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+		// AdditionalProperties informs the schema of any "unknown" keys
+		// Unknown keys are enforced by the ElementType field.
+		if m.AdditionalProperties == nil {
+			// This mirrors how older spec -> proto.Models -> smd would behave.
+			//
+			// proto.Models would unconditionally treat any "object" with empty
+			// "properties"&"additionalProperties" as a Map allowing unknown
+			// keys...I dont think this was explicitly supported behavior.
+			// In Structural Schema docs I've seen no distinction between object
+			// schemas that define properties and those that do not.
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:139]: what is incorrect?

[the remark this answers is not in the document]
Comment (apelisse) [reply pkg/schemaconv/openapi.go:139]: I think objects that don't define properties shouldn't be treated the same as objects that do define them.

[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:139]: Thats fine. The question posed in this comment here is whether an object can have both `Properties` and `AdditionalProperties` at once. The old implementation disallowed this. This new one allows this.

[the remark this answers is not in the document]
Comment (apelisse) [reply pkg/schemaconv/openapi.go:139]: Oh ah yes, we should totally have that! SMD also supports it.

File: pkg/schemaconv/openapi.go [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+	}
+
+	// Structural Schemas produced by kubernetes follow very specific rules which
+	// we can use to infer the SMD type:
+	switch typ {
+	case "":
+		// Some older k8s specs did not include a type for objects. Keep behavior
+		// consistent with old proto.Models implementation:
+		//	If there are Properties, it is a top-level Kind.
+		// 	Otherwise, it is Arbitrary/deduced
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:79]: Im not sure about your idea to treat no type as a deduced type. In the past we've never done this, and reading through eh JSON Schema Core there is no empty `type`. It does note that we are free to extend the type system. To keep with precedent, I would prefer to treat empty the same as `object`, if any thing.

[the remark this answers is not in the document]
Comment (apelisse) [reply pkg/schemaconv/openapi.go:79]: Discussed offline, the proposal doesn't allow less things and seems reasonable I think to implement with the existing type.

File: pkg/schemaconv/testdata/preserve-unknown.yaml [hunk 4]
```diff
@@ -2,31 +2,15 @@ types:
...
-      scalar: untyped
-      list:
-        elementType:
-          namedType: __untyped_atomic_
-        elementRelationship: atomic
-      map:
-        elementType:
-          namedType: __untyped_deduced_
-        elementRelationship: separable
+      namedType: __untyped_deduced_
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/testdata/preserve-unknown.yaml:5]: wasn't sure if it would break other tests.

it did, but I added a leading commit which regenerated expected result `new_schema.yaml` which resolved it. swagger_1.26 is now removed

File: pkg/schemaconv/openapi.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go]: current implementation takes a stab at supporting this any type. please take a look @apelisse

File: pkg/schemaconv/openapi.go [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+				subSchema = &m.Items.Schemas[0]
+			}
+			return c.makeOpenAPIRef(subSchema)
+
+		}
+
+		// A list with no items specified is treated as arbitrary.
+		return schema.TypeRef{
+			NamedType: &untypedName,
+		}
```
Comment (apelisse) [inline pkg/schemaconv/openapi.go:246]: Shouldn't we only allow that if `type` wasn't specified? I feel like if `type: array`, `items` is required

File: pkg/schemaconv/testdata/preserve-unknown.yaml [hunk 6]
```diff
@@ -2,31 +2,15 @@ types:
...
 - name: io.k8s.testcase.EmptyPreserveUnknownFieldsObject
   map:
     fields:
     - name: preserveField
       type:
         map:
```
Comment (apelisse) [inline pkg/schemaconv/testdata/preserve-unknown.yaml:19]: Did we have `map` twice in that? That looks super bogus?

**Review by apelisse — COMMENTED** [review]
That looks really clean and easy to read, thank you!

only minor nits

File: pkg/schemaconv/openapi_test.go [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+// Using all models defined in swagger.json
+// Convert to SMD using two methods:
+//  1. Spec -> SMD
+//  2. Spec -> JSON -> gnostic -> SMD
+//
+// Compare YAML forms. We have some allowed differences...
+func TestOpenAPIImplementation(t *testing.T) {
+	GoldenSchema, err := specToSchemaViaProtoModels([]byte(swaggerJSON))
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi_test.go:60]: I think you dislike it more than I like it xD

File: pkg/schemaconv/testdata/preserve-unknown.yaml (hunk 6 above)
Comment (alexzielenski) [reply pkg/schemaconv/testdata/preserve-unknown.yaml:19]: You are talking about `io.k8s.testcase.EmptyPreserveUnknownFieldsObject`? Looks right to me? Source OpenAPI:

```json
    "io.k8s.testcase.EmptyPreserveUnknownFieldsObject": {
      "description": "",
      "properties": {
        "preserveField": {
          "description": "",
          "type": "object",
          "x-kubernetes-preserve-unknown-fields": true
        }
      }
```

File: pkg/schemaconv/openapi.go [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+		// compute the type inline
+		c2 := c.push("inlined in "+c.currentName, &tr.Inlined)
+		c2.preserveUnknownFields = c.preserveUnknownFields
+		c2.visitSpec(specSchema)
+		c.pop(c2)
+
+		if tr == (schema.TypeRef{}) {
+			// emit warning?
+			tr = schema.TypeRef{
+				NamedType: &untypedName,
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:151]: hmm that was logic copied from protoModels code. the way ive written the openapi version you re right this case is impossible

File: pkg/schemaconv/openapi_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (apelisse) [reply pkg/schemaconv/openapi_test.go:60]: We're not doing Java here, go home!

File: pkg/schemaconv/testdata/preserve-unknown.yaml (hunk 6 above)
Comment (apelisse) [reply pkg/schemaconv/testdata/preserve-unknown.yaml:19]: I don't know if I'm reading the diff well, but it looks like the `io.k8s.testcase.EmptyPreserveUnknownFieldsObject` had two "maps" field, which seems very wrong?

File: pkg/schemaconv/openapi.go (hunk 5 above)
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:246]: Yeah I can make that log an error, but still this function should always return a value. Now that I'm looking at it it seems I have different treatments of each error:

https://github.com/kubernetes/kube-openapi/blob/f065637e0d18a9f594ea3a68afbfc174778ff295/pkg/schemaconv/openapi.go#L230-L232

This case returns a deduced value. I should reorganize this function to avoid mistakes like that.

What should the `items` be inferred to be in case of an error? I vote untyped makes the most sense?

File: pkg/schemaconv/openapi_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/openapi_test.go:60]: reacted to your own comment, eh?

File: pkg/schemaconv/testdata/preserve-unknown.yaml (hunk 6 above)
Comment (alexzielenski) [reply pkg/schemaconv/testdata/preserve-unknown.yaml:19]: I dont see it

File: pkg/schemaconv/openapi.go (hunk 5 above)
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:246]: not sure if `type` is unspecified, should array items be `untyped` or `deduced`?

when I changed it the schema for JSONSChemaPropsOrBool (and a few related others) changed:

```
    "io.k8s.apiextensions-apiserver.pkg.apis.apiextensions.v1.JSONSchemaPropsOrArray": {
      "description": "JSONSchemaPropsOrArray represents a value that can either be a JSONSchemaProps or an array of JSONSchemaProps. Mainly here for serialization purposes."
    },
```

File: pkg/schemaconv/smd.go [hunk 9]
```diff
@@ -355,117 +230,108 @@ func toStringSlice(o interface{}) (out []string, ok bool) {
...
-	ext := m.GetExtensions()
-	if val, ok := ext["x-kubernetes-map-type"]; ok {
+func getListElementRelationship(ext map[string]any) (schema.ElementRelationship, []string, error) {
+	if val, ok := ext["x-kubernetes-list-type"]; ok {
 		switch val {
 		case "atomic":
-			a.Map.ElementRelationship = schema.Atomic
-		case "granular":
-			a.Map.ElementRelationship = schema.Separable
+			return schema.Atomic, nil, nil
```
[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/smd.go:276]: historically they have been ignored

[the remark this answers is not in the document]
Comment (alexzielenski) [reply pkg/schemaconv/smd.go:276]: probably, yes. maybe in followup pr

File: pkg/schemaconv/openapi_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (apelisse) [reply pkg/schemaconv/openapi_test.go]: I've got to pretend I'm funny

File: pkg/schemaconv/testdata/preserve-unknown.yaml (hunk 6 above)
Comment (apelisse) [reply pkg/schemaconv/testdata/preserve-unknown.yaml:19]: Sorry I was confused by the indentation, forget it, it's good (it always was, I was just surprised).

File: pkg/schemaconv/openapi.go [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+)
+
+func ToSchemaFromOpenAPI(models map[string]*spec.Schema, preserveUnknownFields bool) (*schema.Schema, error) {
+	c := convert{
+		preserveUnknownFields: preserveUnknownFields,
+		output:                &schema.Schema{},
+	}
+
+	for name, spec := range models {
+		// Skip/Ignore top-level references
```
Comment (apelisse) [inline pkg/schemaconv/openapi.go:35]: Have you seen some of these, what are they?

**Review by apelisse — COMMENTED** [review]
This is good, but the testing feels mostly limited to built-in types with no tests for CRDs, can we add some CRD tests in there?

File: pkg/schemaconv/openapi_test.go [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return newSchema, nil
+}
+
+func BenchmarkOpenAPIConversion(b *testing.B) {
+	doc := spec.Swagger{}
+	require.NoError(b, doc.UnmarshalJSON([]byte(swaggerJSON)))
```
Comment (apelisse) [inline pkg/schemaconv/openapi_test.go:227]: Should we bench the CRDs too? I feel like that'd be too many, and not sure it would add a lot.

Comment (apelisse) [reply pkg/schemaconv/openapi_test.go:227]: Maybe you don't need ALL the CRDs, just pick a few?

File: pkg/schemaconv/openapi_test.go [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+	return nil
+}
+
+func TestCRDOpenAPIConversion(t *testing.T) {
+	files, err := crdFS.ReadDir("testdata/crds")
+	require.NoError(t, err)
+	for _, entry := range files {
+		t.Run(entry.Name(), func(t *testing.T) {
+			t.Parallel()
+			contents, err := crdFS.ReadFile("testdata/crds/" + entry.Name())
```
Comment (apelisse) [inline pkg/schemaconv/openapi_test.go:181]: I don't know if that's the test I would have expected for CRDs, you're running this against the v2 processed version. But we should compare the `v2 processed version + proto + conversion` against the `v3 + new conversion`, no?

File: pkg/schemaconv/openapi_test.go [hunk 13]
```diff
@@ -0,0 +1,281 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schemaconv_test
+
+import (
+	"embed"
+	_ "embed"
```
Comment (apelisse) [inline pkg/schemaconv/openapi_test.go:21]: hm?

**Review by apelisse — COMMENTED** [review]
Looks generally good. Maybe improve the benchmark a little bit to try some CRDs, and reduce the number of CRDs now that we know this works reasonably well.

**Comment by alexzielenski** [comment]
reduced # crds to 10. added benchmark. results very promising:

```console
goos: darwin
goarch: arm64
pkg: k8s.io/kube-openapi/pkg/schemaconv
BenchmarkOpenAPICRDConversion
BenchmarkOpenAPICRDConversion/accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json
BenchmarkOpenAPICRDConversion/accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json/spec.Schema->schema.Schema-20                 14871             80875 ns/op           85328 B/op       1102 allocs/op
BenchmarkOpenAPICRDConversion/accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20               100          10065875 ns/op         6153759 B/op      51982 allocs/op
BenchmarkOpenAPICRDConversion/container_v1beta1_containercluster.json
BenchmarkOpenAPICRDConversion/container_v1beta1_containercluster.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/container_v1beta1_containercluster.json/spec.Schema->schema.Schema-20                                                                             9154            129187 ns/op          120010 B/op       1674 allocs/op
BenchmarkOpenAPICRDConversion/container_v1beta1_containercluster.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/container_v1beta1_containercluster.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                              100          10791915 ns/op         6712496 B/op      60459 allocs/op
BenchmarkOpenAPICRDConversion/dataproc_v1beta1_dataprocworkflowtemplate.json
BenchmarkOpenAPICRDConversion/dataproc_v1beta1_dataprocworkflowtemplate.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/dataproc_v1beta1_dataprocworkflowtemplate.json/spec.Schema->schema.Schema-20                                                                      8398            138465 ns/op          136810 B/op       1801 allocs/op
BenchmarkOpenAPICRDConversion/dataproc_v1beta1_dataprocworkflowtemplate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/dataproc_v1beta1_dataprocworkflowtemplate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                       100          12267150 ns/op         7909691 B/op      63534 allocs/op
BenchmarkOpenAPICRDConversion/dlp_v1beta1_dlpdeidentifytemplate.json
BenchmarkOpenAPICRDConversion/dlp_v1beta1_dlpdeidentifytemplate.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/dlp_v1beta1_dlpdeidentifytemplate.json/spec.Schema->schema.Schema-20                                                                              4594            266341 ns/op          273684 B/op       3253 allocs/op
BenchmarkOpenAPICRDConversion/dlp_v1beta1_dlpdeidentifytemplate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/dlp_v1beta1_dlpdeidentifytemplate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                                62          18247396 ns/op        11624966 B/op      85813 allocs/op
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringalertpolicy.json
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringalertpolicy.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringalertpolicy.json/spec.Schema->schema.Schema-20                                                                      27733             42827 ns/op           43840 B/op        627 allocs/op
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringalertpolicy.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringalertpolicy.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                        135           8838077 ns/op         5586933 B/op      46326 allocs/op
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringdashboard.json
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringdashboard.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringdashboard.json/spec.Schema->schema.Schema-20                                                                         5194            232198 ns/op          252571 B/op       3126 allocs/op
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringdashboard.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/monitoring_v1beta1_monitoringdashboard.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                           45          25979421 ns/op        17066856 B/op      99472 allocs/op
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificate.json
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificate.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificate.json/spec.Schema->schema.Schema-20                                                                        15192             79110 ns/op           82625 B/op       1106 allocs/op
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificate.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                          130           9163183 ns/op         5806177 B/op      51355 allocs/op
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificateauthority.json
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificateauthority.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificateauthority.json/spec.Schema->schema.Schema-20                                                               13740             86605 ns/op           90617 B/op       1184 allocs/op
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificateauthority.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/privateca_v1beta1_privatecacertificateauthority.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                 124           9552285 ns/op         5966763 B/op      52560 allocs/op
BenchmarkOpenAPICRDConversion/storagetransfer_v1beta1_storagetransferjob.json
BenchmarkOpenAPICRDConversion/storagetransfer_v1beta1_storagetransferjob.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/storagetransfer_v1beta1_storagetransferjob.json/spec.Schema->schema.Schema-20                                                                    23853             49901 ns/op           50696 B/op        712 allocs/op
BenchmarkOpenAPICRDConversion/storagetransfer_v1beta1_storagetransferjob.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/storagetransfer_v1beta1_storagetransferjob.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                      146           8209869 ns/op         5240361 B/op      46030 allocs/op
BenchmarkOpenAPICRDConversion/vpcaccess_v1beta1_vpcaccessconnector.json
BenchmarkOpenAPICRDConversion/vpcaccess_v1beta1_vpcaccessconnector.json/spec.Schema->schema.Schema
BenchmarkOpenAPICRDConversion/vpcaccess_v1beta1_vpcaccessconnector.json/spec.Schema->schema.Schema-20                                                                          41829             28973 ns/op           28472 B/op        420 allocs/op
BenchmarkOpenAPICRDConversion/vpcaccess_v1beta1_vpcaccessconnector.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema
BenchmarkOpenAPICRDConversion/vpcaccess_v1beta1_vpcaccessconnector.json/spec.Schema->json->gnostic_v2->proto.Models->schema.Schema-20                                            162           7501460 ns/op         4714945 B/op      41656 allocs/op
PASS
ok      k8s.io/kube-openapi/pkg/schemaconv      32.063s
```

File: pkg/schemaconv/openapi.go [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+import (
+	"errors"
+	"path"
+	"strings"
+
+	"k8s.io/kube-openapi/pkg/validation/spec"
+	"sigs.k8s.io/structured-merge-diff/v4/schema"
+)
+
+func ToSchemaFromOpenAPI(models map[string]*spec.Schema, preserveUnknownFields bool) (*schema.Schema, error) {
```
Comment (apelisse) [inline pkg/schemaconv/openapi.go:28]: I would expect this function to be the only new part of the interface? Can we document that method? Are there any restrictions on the `spec.Schema`? Do they have to be v2, v3, or anything else? Any restrictions? Should they be structural? etc

File: pkg/builder3/openapi.go [hunk 15]
```diff
@@ -327,6 +327,21 @@ func BuildOpenAPISpecFromRoutes(webServices []common.RouteContainer, config *com
 	return a.spec, nil
 }
 
+// BuildOpenAPIDefinitionsForResource builds a partial OpenAPI spec given a sample object and common.Config to customize it.
+// BuildOpenAPIDefinitionsForResources returns the OpenAPI spec which includes the definitions for the
+// passed type names.
+func BuildOpenAPIDefinitionsForResources(config *common.Config, names ...string) (map[string]*spec.Schema, error) {
```
Comment (apelisse) [inline pkg/builder3/openapi.go:333]: Why did we add this? Is this related to the PR, or is this something else?

**Review by apelisse — COMMENTED** [review]
Looks good. I don't want to block on the comment, it can be added in another PR (please don't forget)

The performance will be even better than highlighted for a few different reasons I think:
1. We don't have to merge the specs the way we do today
2. This already assumes we have made v3 -> v2, so it should be even faster
Right?

**Comment by apelisse** [comment]
/lgtm
/approve

Not blocking for comment and question, thanks

File: pkg/builder3/openapi.go (hunk 15 above)
Comment (alexzielenski) [reply pkg/builder3/openapi.go:333]: ah its used by the integration

File: pkg/schemaconv/openapi.go (hunk 14 above)
Comment (alexzielenski) [reply pkg/schemaconv/openapi.go:28]: We expect a V3 structural schema - same requirements for `openAPIV3Schema` field of CRDs. Will make doc pr

Comment (apelisse) [reply pkg/schemaconv/openapi.go:28]: Thanks!

**Comment by apelisse** [comment]
Reformatting benchmarks for clarity:
```
name                                                                                            old time/op    new time/op    delta
accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json-20    10.1ms ± 0%     0.1ms ± 0%  -99.20%  (p=0.008 n=5+5)
container_v1beta1_containercluster.json-20                                   10.8ms ± 0%     0.1ms ± 0%  -98.80%  (p=0.008 n=5+5)
dataproc_v1beta1_dataprocworkflowtemplate.json-20                            12.3ms ± 0%     0.1ms ± 0%  -98.87%  (p=0.008 n=5+5)
dlp_v1beta1_dlpdeidentifytemplate.json-20                                    18.2ms ± 0%     0.3ms ± 0%  -98.54%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringalertpolicy.json-20                             8.84ms ± 0%    0.04ms ± 0%  -99.52%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringdashboard.json-20                               26.0ms ± 0%     0.2ms ± 0%  -99.11%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificate.json-20                               9.16ms ± 0%    0.08ms ± 0%  -99.14%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificateauthority.json-20                      9.55ms ± 0%    0.09ms ± 0%  -99.09%  (p=0.008 n=5+5)
storagetransfer_v1beta1_storagetransferjob.json-20                           8.21ms ± 0%    0.05ms ± 0%  -99.39%  (p=0.008 n=5+5)
vpcaccess_v1beta1_vpcaccessconnector.json-20                                 7.50ms ± 0%    0.03ms ± 0%  -99.61%  (p=0.008 n=5+5)

name                                                                                            old alloc/op   new alloc/op   delta
accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json-20    6.15MB ± 0%    0.09MB ± 0%  -98.61%  (p=0.008 n=5+5)
container_v1beta1_containercluster.json-20                                   6.71MB ± 0%    0.12MB ± 0%  -98.21%  (p=0.008 n=5+5)
dataproc_v1beta1_dataprocworkflowtemplate.json-20                            7.91MB ± 0%    0.14MB ± 0%  -98.27%  (p=0.008 n=5+5)
dlp_v1beta1_dlpdeidentifytemplate.json-20                                    11.6MB ± 0%     0.3MB ± 0%  -97.65%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringalertpolicy.json-20                             5.59MB ± 0%    0.04MB ± 0%  -99.22%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringdashboard.json-20                               17.1MB ± 0%     0.3MB ± 0%  -98.52%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificate.json-20                               5.81MB ± 0%    0.08MB ± 0%  -98.58%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificateauthority.json-20                      5.97MB ± 0%    0.09MB ± 0%  -98.48%  (p=0.008 n=5+5)
storagetransfer_v1beta1_storagetransferjob.json-20                           5.24MB ± 0%    0.05MB ± 0%  -99.03%  (p=0.008 n=5+5)
vpcaccess_v1beta1_vpcaccessconnector.json-20                                 4.71MB ± 0%    0.03MB ± 0%  -99.40%  (p=0.008 n=5+5)

name                                                                                            old allocs/op  new allocs/op  delta
accesscontextmanager_v1beta1_accesscontextmanagerserviceperimeter.json-20     52.0k ± 0%      1.1k ± 0%  -97.88%  (p=0.008 n=5+5)
container_v1beta1_containercluster.json-20                                    60.5k ± 0%      1.7k ± 0%  -97.23%  (p=0.008 n=5+5)
dataproc_v1beta1_dataprocworkflowtemplate.json-20                             63.5k ± 0%      1.8k ± 0%  -97.17%  (p=0.008 n=5+5)
dlp_v1beta1_dlpdeidentifytemplate.json-20                                     85.8k ± 0%      3.3k ± 0%  -96.21%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringalertpolicy.json-20                              46.3k ± 0%      0.6k ± 0%  -98.65%  (p=0.008 n=5+5)
monitoring_v1beta1_monitoringdashboard.json-20                                99.5k ± 0%      3.1k ± 0%  -96.86%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificate.json-20                                51.4k ± 0%      1.1k ± 0%  -97.85%  (p=0.008 n=5+5)
privateca_v1beta1_privatecacertificateauthority.json-20                       52.6k ± 0%      1.2k ± 0%  -97.75%  (p=0.008 n=5+5)
storagetransfer_v1beta1_storagetransferjob.json-20                            46.0k ± 0%      0.7k ± 0%  -98.45%  (p=0.008 n=5+5)
vpcaccess_v1beta1_vpcaccessconnector.json-20                                  41.7k ± 0%      0.4k ± 0%  -98.99%  (p=0.008 n=5+5)
```

**Comment by lavalamp** [comment]
Two orders of magnitude, nice!
